### PR TITLE
Lessen constraints on certain algorithm arguments

### PR DESCRIPTION
**Pull request by hkaiser**
- some sequential algorithms (copy, etc.) are fully functional if - when used sequentially - the destination is just an output iterator
- fixing is_output_iterator
- fixing `transform_reduce` implementation to properly handle constant input sequences
- flyby: minor formatting changes

### CODE REVIEW

File: libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform_reduce.hpp [hunk 1]
```diff
@@ -625,16 +626,16 @@ namespace hpx { namespace ranges {
...
                 >::value &&
                 hpx::traits::is_invocable<Reduce,
                     typename hpx::util::invoke_result<Convert,
-                        typename hpx::traits::range_iterator<Rng>::type,
+                        typename hpx::traits::range_traits<Rng>::value_type,
                         typename std::iterator_traits<Iter2>::value_type
                     >::type,
                     typename hpx::util::invoke_result<Convert,
-                        typename hpx::traits::range_iterator<Rng>::type,
+                        typename hpx::traits::range_traits<Rng>::value_typee,
```
Comment (msimberg) [inline libs/parallelism/algorithms/include/hpx/parallel/container_algorithms/transform_reduce.hpp:638]:
```suggestion
                        typename hpx::traits::range_traits<Rng>::value_type,
```

**Review by msimberg — CHANGES_REQUESTED** [review]
Thanks @hkaiser, looks good to me.

**Review by aurianer — APPROVED** [review]
LGTM thanks!

**Comment by hkaiser** [comment]
@msimberg I think is this is ready to go.

**Comment by msimberg** [comment]
retest

**Pull request merged**